todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/addedittask/AddEditTaskFragment.java

```diff
@@ -16,101 +16,166 @@
 import android.support.design.widget.Snackbar;
 import android.support.v4.app.Fragment;
+import android.util.Log;
 import android.view.LayoutInflater;
 import android.view.View;
 import android.view.ViewGroup;
 import android.widget.TextView;
 
+import com.example.android.architecture.blueprints.todoapp.Injection;
 import com.example.android.architecture.blueprints.todoapp.R;
+import com.example.android.architecture.blueprints.todoapp.data.Task;
 
-import static com.google.common.base.Preconditions.checkNotNull;
+import rx.android.schedulers.AndroidSchedulers;
+import rx.schedulers.Schedulers;
+import rx.subscriptions.CompositeSubscription;
 
 /**
  * Main UI for the add task screen. Users can enter a task title and description.
  */
-public class AddEditTaskFragment extends Fragment implements AddEditTaskContract.View {
+public class AddEditTaskFragment extends Fragment {
 
     public static final String ARGUMENT_EDIT_TASK_ID = "EDIT_TASK_ID";
-
-    private AddEditTaskContract.Presenter mPresenter;
+    private static final String TASK_TITLE_KEY = "title";
+    private static final String TASK_DESCRIPTION_KEY = "description";
+    private static final String TAG = AddEditTaskFragment.class.getSimpleName();
 
     private TextView mTitle;
 
     private TextView mDescription;
 
-    public static AddEditTaskFragment newInstance() {
-        return new AddEditTaskFragment();
+    private AddEditTaskViewModel mViewModel;
+
+    private CompositeSubscription mSubscription = new CompositeSubscription();
+
+    public static AddEditTaskFragment newInstance(String taskId) {
+        Bundle arguments = new Bundle();
+        arguments.putString(ARGUMENT_EDIT_TASK_ID, taskId);
+        AddEditTaskFragment fragment = new AddEditTaskFragment();
+        fragment.setArguments(arguments);
+        return fragment;
+    }
+
+    @Nullable
+    @Override
+    public View onCreateView(LayoutInflater inflater, ViewGroup container,
+                             Bundle savedInstanceState) {
+        View root = inflater.inflate(R.layout.addtask_frag, container, false);
+        mTitle = (TextView) root.findViewById(R.id.add_task_title);
+        mDescription = (TextView) root.findViewById(R.id.add_task_description);
+        setHasOptionsMenu(true);
+
+        setupFab();
+
+        mViewModel = Injection.createAddEditTaskViewModel(getTaskId(), getActivity());
+        restoreData(savedInstanceState);
+
+        return root;
     }
 
     @Override
     public void onResume() {
+        bindViewModel();
         super.onResume();
-        mPresenter.subscribe();
     }
 
     @Override
     public void onPause() {
+        unbindViewModel();
         super.onPause();
-        mPresenter.unsubscribe();
     }
 
-    @Override
-    public void setPresenter(@NonNull AddEditTaskContract.Presenter presenter) {
-        mPresenter = checkNotNull(presenter);
+    private void bindViewModel() {
+        mSubscription = new CompositeSubscription();
```

Comment:
// Subscriptions are gathered in a CompositeSubscription to be able to unsubsc....